src/main/java/org/folio/bulkops/service/QueryService.java

```diff
@@ -1,94 +1,318 @@
+      var matchedCsvFileName = operation.getId() + SLASH + LocalDate.now() + MATCHED_RECORDS + FilenameUtils.getBaseName(operation.getLinkToTriggeringCsvFile()) + ".csv";
+
+      operation.setStatus(RETRIEVING_RECORDS);
+      bulkOperationRepository.save(operation);
+
+      processQueryResult(queryResult, matchedCsvFileName, matchedJsonFileName, matchedMrcFileName, operation);
+
+      if (operation.getMatchedNumOfRecords() > 0) {
+        operation.setLinkToMatchedRecordsCsvFile(matchedCsvFileName);
+        operation.setLinkToMatchedRecordsJsonFile(matchedJsonFileName);
+        operation.setStatus(DATA_MODIFICATION);
+      } else {
+        operation.setStatus(COMPLETED_WITH_ERRORS);
+      }
+      operation.setEndTime(LocalDateTime.now());
+      bulkOperationRepository.save(operation);
+    } catch (Exception e) {
+      log.error(ERROR_STARTING_BULK_OPERATION, e);
+      operation.setStatus(FAILED);
+      operation.setErrorMessage(e.getMessage());
+      operation.setEndTime(LocalDateTime.now());
+      var linkToMatchingErrorsFile = errorService.uploadErrorsToStorage(operation.getId(), ERROR_MATCHING_FILE_NAME_PREFIX, e.getMessage());
+      operation.setLinkToMatchedRecordsErrorsCsvFile(linkToMatchingErrorsFile);
+    }
+  }
+
+  private void processQueryResult(QueryDetails queryDetails, String matchedCsvFileName, String matchedJsonFileName,
+                                  String matchedMrcFileName, BulkOperation operation) throws IOException, CsvRequiredFieldEmptyException, CsvDataTypeMismatchException {
+    try (var writerForResultCsvFile = remoteFileSystemClient.writer(matchedCsvFileName);
+         var writerForResultJsonFile = remoteFileSystemClient.writer(matchedJsonFileName);
+         var writerForResultMrcFile = remoteFileSystemClient.writer(matchedMrcFileName)) {
+      var entityClass = resolveEntityClass(operation.getEntityType());
+      var csvWriter = new BulkOperationsEntityCsvWriter(writerForResultCsvFile, entityClass);
+      List<BulkOperationExecutionContent> bulkOperationExecutionContents = new ArrayList<>();
+
+      int numMatched = 0;
+      int numProcessed = 0;
+      for (var content : queryDetails.getContent()) {
+
+        var entityRecord = constructRecord(operation, content);
+
+        try {
+          permissionsValidator.checkPermissions(operation, entityRecord);
+
+          var extendedRecord = constructExtendedRecord(operation, entityRecord);
+
+          if (entityRecord.isMarcInstance()) {
+            processQueryResultForMarc(entityRecord, writerForResultMrcFile, operation, matchedMrcFileName);
+          }
+
+          var extendedRecordAsJsonString = objectMapper.writeValueAsString(extendedRecord);
+
+          writerForResultJsonFile.append(extendedRecordAsJsonString);
+          CSVHelper.writeBeanToCsv(operation, csvWriter, extendedRecord.getRecordBulkOperationEntity(), bulkOperationExecutionContents);
+          numMatched++;
+        } catch (UploadFromQueryException e) {
```

Comment:
What will happen if CSVHelper.writeBeanToCsv throws exception from some converter which comes from particular service. For example, some value cannot be resolved by id? From the first look error which is just warning and should affect only one entity will be rethrown above and will led to bulk operation fail?